sdks/java/core/src/main/java/org/apache/beam/sdk/util/ApiSurface.java

```diff
@@ -66,45 +80,283 @@
+                return checkedApiSurface.getAnyExposurePath(aClass);
+              }
+            };
+
+        final Maps.EntryTransformer<Class<?>, List<Class<?>>, String> toMessage =
+            new Maps.EntryTransformer<Class<?>, List<Class<?>>, String>() {
+
+              @Override
+              public String transformEntry(
+                  @Nonnull final Class<?> aClass, @Nonnull final List<Class<?>> exposure) {
+                return aClass + " exposed via:\n\t\t" + Joiner.on("\n\t\t").join(exposure);
+              }
+            };
+
+        final Predicate<Class<?>> disallowed =
+            new Predicate<Class<?>>() {
+
+              @Override
+              public boolean apply(@Nonnull final Class<?> aClass) {
+                return !classIsAllowed(aClass, allowedClasses);
+              }
+            };
+
+        /* </helper_lambdas> */
+
+        final FluentIterable<Class<?>> disallowedClasses =
+            FluentIterable.from(checkedApiSurface.getExposedClasses()).filter(disallowed);
+
+        final ImmutableMap<Class<?>, List<Class<?>>> exposures =
+            Maps.toMap(disallowedClasses, toExposure);
+
+        final ImmutableList<String> messages =
+            FluentIterable.from(Maps.transformEntries(exposures, toMessage).values())
+                .toSortedList(Ordering.<String>natural());
+
+        if (!messages.isEmpty()) {
+          mismatchDescription.appendText(
+              "The following disallowed classes appeared on the API surface:\n\t"
+                  + Joiner.on("\n\t").join(messages));
+        }
+
+        return messages.isEmpty();
+      }
+
+      @SuppressWarnings({"rawtypes", "unchecked"})
+      private boolean classIsAllowed(
+          final Class<?> clazz, final Set<Matcher<Class<?>>> allowedClasses) {
+        // Safe cast inexpressible in Java without rawtypes
+        return anyOf((Iterable) allowedClasses).matches(clazz);
+      }
+
+      @Override
+      protected boolean matchesSafely(
+          final ApiSurface apiSurface, final Description mismatchDescription) {
+        final boolean noDisallowed =
+            verifyNoDisallowed(apiSurface, classMatchers, mismatchDescription);
+
+        final boolean noAbandoned =
+            verifyNoAbandoned(apiSurface, classMatchers, mismatchDescription);
+
+        return noDisallowed & noAbandoned;
+      }
+
+      @Override
+      public void describeTo(final Description description) {
+        description.appendText("API surface to include only:" + "\n\t");
+        for (final Matcher<Class<?>> classMatcher : classMatchers) {
+          classMatcher.describeTo(description);
+          description.appendText("\n\t");
+        }
+      }
+    }
+  }
+
+  ///////////////
+
+  private ClassLoader getSdkClassLoader() {
+    final ClassLoader myClassLoader = getClass().getClassLoader();
+    final ClassLoader systemClassLoader = ClassLoader.getSystemClassLoader();
+    return systemClassLoader.equals(myClassLoader)
```

Comment:
That's a good point, it appears I got a bit over-zealous to try and preserve previous behavior :)

I think that making the `ClassLoader` instance a dependency that is passed in by the caller makes a lot of sense.